File: library/src/test/java/io/github/anderscheow/validator/rules/common/MaxRuleTest.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.StringRes;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Locale;

import static org.junit.Assert.*;

public class MaxRuleTest {

    private static final int MAX_LENGTH = 5;

    private MaxRule maxRule;

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
        maxRule = null;
    }

    @Test(expected = NullPointerException.class)
    public void validate_EmptySample_ThrowNullPointerException() throws Exception {
        maxRule = new MaxRule(MAX_LENGTH);

        maxRule.validate(null);
    }

    @Test
    public void validate_SampleEqualMax_ReturnTrue() throws Exception {
        maxRule = new MaxRule(MAX_LENGTH);

        String sample = "tests";

        assertTrue(maxRule.validate(sample));
    }

    @Test
    public void validate_ValidSampleLessThanMax_ReturnTrue() throws Exception {
        maxRule = new MaxRule(MAX_LENGTH);

        String sample = "test";

        assertTrue(maxRule.validate(sample));
    }

    @Test
    public void validate_SampleMoreThanMax_ReturnTrue() throws Exception {
        maxRule = new MaxRule(MAX_LENGTH);

        String sample = "testing";

        assertFalse(maxRule.validate(sample));
    }

    @Test
    public void errorMessage_DefaultErrorMessage() throws Exception {
        String errorMessage = String.format(Locale.getDefault(), "Length must not exceed %d characters", MAX_LENGTH);

        maxRule = new MaxRule(MAX_LENGTH);

        assertEquals(errorMessage, maxRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorMessage() throws Exception {
        String errorMessage = "This is a custom error message";

        maxRule = new MaxRule(MAX_LENGTH, errorMessage);

        assertEquals(errorMessage, maxRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorRes() throws Exception {
        @StringRes int errorRes = 0;

        maxRule = new MaxRule(MAX_LENGTH, errorRes);

        assertEquals(errorRes, maxRule.errorRes());
    }
}

File: library/src/test/java/io/github/anderscheow/validator/rules/common/MaxRuleTest.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class MaxRuleTest {

    private lateinit var maxRule: MaxRule

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test(expected = NullPointerException::class)
    @Throws(Exception::class)
    fun validate_EmptySample_ThrowNullPointerException() {
        maxRule = MaxRule(MAX_LENGTH, "")

        maxRule.validate(null)
    }

    @Test
    @Throws(Exception::class)
    fun validate_StringSampleEqualMax_ReturnTrue() {
        maxRule = MaxRule(MAX_LENGTH, "")

        val sample = "test"

        assertTrue(maxRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun validate_StringSampleLessThanMax_ReturnTrue() {
        maxRule = MaxRule(MAX_LENGTH, "")

        val sample = "tes"

        assertTrue(maxRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun validate_SampleMoreThanMax_ReturnTrue() {
        maxRule = MaxRule(MAX_LENGTH, "")

        val sample = "testing"

        assertFalse(maxRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorMessage() {
        val errorMessage = "This is a custom error message"

        maxRule = MaxRule(MAX_LENGTH, errorMessage)

        assertEquals(errorMessage, maxRule.errorString)
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorRes() {
        @StringRes val errorRes = 0

        maxRule = MaxRule(MAX_LENGTH, errorRes)

        assertEquals(errorRes, maxRule.errorRes)
    }

    companion object {
        private const val MAX_LENGTH = 4
    }
}